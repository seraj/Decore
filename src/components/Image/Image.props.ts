import ThemeProps from "../../theme/Theme.props";
import { Theme } from "../../theme";

export default interface ImageProps
  extends ThemeProps<
    React.ImgHTMLAttributes<HTMLImageElement>,
    HTMLImageElement
  > {
  lazy?: boolean;

  theme: Theme;
}
