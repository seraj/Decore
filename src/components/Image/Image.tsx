import * as React from "react";
import LazyLoad from "react-lazyload";
import { CSSTransition } from "react-transition-group";
import { withTheme } from "styled-components";
import queryString from "query-string";
import StyledImage from "./styled/StyledImage";
import ImageProps from "./Image.props";

interface ImageState {
  loaded: boolean;
}
class Image extends React.Component<ImageProps, ImageState> {
  public state = {
    loaded: false
  };

  private onLoad = () => {
    this.setState({
      loaded: true
    });
  };

  private handleImageSource = (
    height: string | number | undefined,
    width: string | number | undefined,
    resize: boolean | undefined,
    src: string | undefined
  ) => {
    let source;
    const parseUrl = queryString.parseUrl(src || "");
    if (height !== undefined && width !== undefined && resize) {
      const query = queryString.stringify({
        ...parseUrl.query,
        resize: `${width}x${height}`
      });
      source = `${parseUrl.url}?${query}`;
    } else {
      source = src;
    }
    return source;
  };
  public render() {
    const {
      props: { lazy, theme, src, ...props },
      state: { loaded }
    } = this;
    const source = this.handleImageSource(
      props.height,
      props.width,
      props.resize,
      src
    );
    const image = <StyledImage src={source} {...props} onLoad={this.onLoad} />;

    if (lazy) {
      return (
        <LazyLoad height={props.height}>
          <CSSTransition
            timeout={theme.animations.duration}
            in={loaded}
            classNames="decore-image"
          >
            {image}
          </CSSTransition>
        </LazyLoad>
      );
    } else {
      return image;
    }
  }
}

export default withTheme(Image);
