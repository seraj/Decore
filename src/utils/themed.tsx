import * as React from "react";
import * as renderer from "react-test-renderer";
import Decore from "../components/Decore";
import theme from "../theme";

const themed = (component: JSX.Element) =>
  renderer.create(<Decore theme={theme}>{component}</Decore>);

export default themed;
