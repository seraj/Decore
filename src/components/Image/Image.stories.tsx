import { text, withKnobs } from "@storybook/addon-knobs";
import { storiesOf } from "@storybook/react";
import * as React from "react";
import wInfo from "../../utils/wInfo";
import Image from "./Image";

const stories = storiesOf("Components/Image", module);
stories.addDecorator(withKnobs);

stories.add(
  "default",
  wInfo()(() => (
    <div>
      <Image
        lazy={true}
        src={`https://dummyimage.com/200x600/000/fff.jpg?text=${Date.now()}`}
        height={640}
      />
      <br />
      <Image
        lazy={true}
        src={`https://dummyimage.com/200x600/000/fff.jpg?text=${Date.now() +
          1}`}
        height={640}
      />
      <br />
      <Image
        lazy={true}
        src={`https://dummyimage.com/200x600/000/fff.jpg?text=${Date.now() +
          2}`}
        height={640}
      />
      <br />
      <Image
        lazy={true}
        src={`https://dummyimage.com/200x600/000/fff.jpg?text=${Date.now() +
          3}`}
        height={640}
      />
      <br />
    </div>
  ))
);
