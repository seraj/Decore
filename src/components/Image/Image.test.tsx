import * as React from "react";

import themed from "../../utils/themed";
import Image from "./Image";

describe("Image", () => {
  it.skip("should render correctly", () => {
    const component = themed(<Image src="test" />).toJSON();

    expect(component).toMatchSnapshot();
  });
});
