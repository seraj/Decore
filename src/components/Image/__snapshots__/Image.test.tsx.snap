// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should render correctly 1`] = `
.c0.decore-image-enter {
  opacity: 0;
}

.c0.decore-image-enter-active {
  opacity: 1;
  -webkit-transition: opacity 250ms;
  transition: opacity 250ms;
}

.c0.decore-image-exit {
  opacity: 1;
}

.c0.decore-image-exit {
  opacity: 0;
  -webkit-transition: opacity 250ms;
  transition: opacity 250ms;
}

<img
  className="c0"
  onLoad={[Function]}
>
  Test
</img>
`;
